### DISCUSSION TITLE
Help in generating extent reports for parallel execution!

### DISCUSSION THREAD

**Discussion by Balakumar-14** [appium/appium#17329] [Q&A]
I am a beginner at appium parallel execution. And now, I am stuck with generating reports for parallel execution. 
**_Case: If 2 cases run parallelly, both cases run parallelly in devices. But, the report of one case is overwriting the other case. So, one case' logs are generated in the report._**
What are the changes I have to do with my code to achieve the expected report?
Code attached below for reference:

Extent report class:
`public class ExtentReportUtil
{
    public static ExtentSparkReporter sparkReporter;
    public static ExtentReports extent;
    public static ExtentTest extTest;
    public static InetAddress ip;
    public static String hostName;
    public static String userName;
    public static String sparkReportFile;

    public static Logger log4j = LogManager.getLogger(ExtentReportUtil.class);

    public static void createSparkReport()
    {
        SimpleDateFormat format = new SimpleDateFormat("dd-MM-yy hh:mm:ss aa");
        Date date = new Date();
        String dateStr = format.format(date);
        sparkReportFile = System.getProperty("user.dir") + "/Reports/SparkReports/SparkReport_" + dateStr + ".html";

        sparkReporter = new ExtentSparkReporter(sparkReportFile)
                .viewConfigurer()
                .viewOrder()
                .as(new ViewName[] { ViewName.CATEGORY, ViewName.TEST, ViewName.EXCEPTION , ViewName.DASHBOARD })
                .apply();

        extent = new ExtentReports();
        extent.attachReporter(sparkReporter);

        try
        {
            ip = InetAddress.getLocalHost();
            hostName = ip.getHostName();
            userName = System.getProperty("user.name");
            extent.setSystemInfo("OS",System.getProperty("os.name"));
            extent.setSystemInfo("OS Version",System.getProperty("os.version"));
            extent.setSystemInfo("OS Architecture", System.getProperty("os.arch"));
            extent.setSystemInfo("Java Version : ", System.getProperty("java.version"));
        }
        catch (UnknownHostException ex)
        {
        
        }

        extent.setSystemInfo("Host Name", hostName);
        extent.setSystemInfo("User Name", userName);

        sparkReporter.config().setDocumentTitle(" Spark Report");
        sparkReporter.config().setReportName("Android Regression Automation Test Report");
        sparkReporter.config().setTheme(Theme.DARK);
        sparkReporter.config().setTimelineEnabled(true);
    }

    public static void closeSparkReport()
    {
        extent.flush();
        try
        {
            FileUtils.copyFile(FileUtils.getFile(sparkReportFile),
                    new File(System.getProperty("user.dir") + "/Reports/PublishableReport/index.html"));
        }
        catch (IOException e)
        {
            throw new RuntimeException(e);
        }
    }

    public void createTest(String testName)
    {
        extTest = extent.createTest(testName);
    }
    public void createTestCategory(String categoryName) {
        extTest = extTest.assignCategory(categoryName);
    }

    public void infoWithColor(String msg, ExtentColor colour)
    {
        extTest.log(Status.INFO, MarkupHelper.createLabel(msg, colour));
        log4j.info(msg);
    }

    public void info(String msg)
    {
        extTest.log(Status.INFO, msg);
        log4j.info(msg);
    }

    public void warn(String msg)
    {
        extTest.log(Status.WARNING, MarkupHelper.createLabel(msg, ExtentColor.ORANGE));
        log4j.warn(msg);
    }

    public void fail(String msg)
    {
        extTest.log(Status.FAIL, MarkupHelper.createLabel(msg, ExtentColor.RED));
        log4j.error(msg);
        Assert.fail(msg);
    }

    public void pass(String msg)
    {
        extTest.log(Status.PASS, MarkupHelper.createLabel(msg, ExtentColor.GREEN));
        log4j.info(msg);
    }

    public void fail(Throwable t)
    {
        extTest.fail(t);
    }

    public void getResult(ITestResult result)
    {
        if (result.getStatus() == ITestResult.FAILURE)
        {
            String fileName = composeTestName(result);
            String screenShotPath = System.getProperty("user.dir") + "/ScreenShots/" + fileName + ".PNG";

            ExtentReportUtil.extTest.log(Status.FAIL, MarkupHelper.createLabel(result.getName() + " Test case FAILED due to below issues:", ExtentColor.RED));
            ExtentReportUtil.extTest.fail(result.getThrowable());

            try
            {
                ExtentReportUtil.extTest.fail("Test Case Failed Screenshot" , MediaEntityBuilder.createScreenCaptureFromBase64String(Helper.takeScreenShot(screenShotPath)).build());
            }
            catch (Exception e)
            {ExtentReportUtil.extTest.fail("Screenshot path is wrong and error is " + e.getLocalizedMessage());}

        }
        else if (result.getStatus() == ITestResult.SUCCESS)
        {
            ExtentReportUtil.extTest.log(Status.PASS,
                    MarkupHelper.createLabel(result.getName() + " Test Case PASSED", ExtentColor.GREEN));
        }
        else if (result.getStatus() == ITestResult.SKIP)
        {
            String fileName = composeTestName(result);
            String screenShotPath = System.getProperty("user.dir") + "/ScreenShots/" + fileName + ".png";

            ExtentReportUtil.extTest.log(Status.SKIP, MarkupHelper.createLabel(result.getName() + " Test Case SKIPPED due to below issues ", ExtentColor.ORANGE));
            ExtentReportUtil.extTest.skip(result.getThrowable());

            try
            {
                ExtentReportUtil.extTest.skip("Test Case Skipped Screenshot" , MediaEntityBuilder.createScreenCaptureFromBase64String(Helper.takeScreenShot(screenShotPath)).build());
            }
            catch (Exception e)
            {ExtentReportUtil.extTest.skip("Screenshot path is wrong and error is " + e.getLocalizedMessage());}

        }
        else
        {
            String fileName = composeTestName(result);
            String screenShotPath = System.getProperty("user.dir") + "/ScreenShots/" + fileName + ".png";

            ExtentReportUtil.extTest.log(Status.SKIP, MarkupHelper.createLabel(result.getName() + " Test Case neither PASSED / FAILED / SKIPPED and issue is ", ExtentColor.ORANGE));
            ExtentReportUtil.extTest.skip(result.getThrowable());

            try
            {
                ExtentReportUtil.extTest.skip("Test Case  Screenshot " , MediaEntityBuilder.createScreenCaptureFromBase64String(Helper.takeScreenShot(screenShotPath)).build());
            }
            catch (Exception e)
            {ExtentReportUtil.extTest.fail("Screenshot path is wrong and error is " + e);}

        }
    }
    private String composeTestName(ITestResult iTestResult)
    {
        StringBuilder completeFileName = new StringBuilder();
        completeFileName.append(iTestResult.getTestClass().getRealClass().getSimpleName()); // simplified class name
        completeFileName.append("_");
        completeFileName.append(iTestResult.getName()); // method name

        // all the parameters information
        Object[] parameters = iTestResult.getParameters();
        for (Object parameter : parameters)
        {
            completeFileName.append("_");
            completeFileName.append(parameter);
        }

        SimpleDateFormat format = new SimpleDateFormat("dd-MM-yy hh:mm:ss aa");
        Date date = new Date();
        String dateStr = format.format(date);

        String temp = completeFileName.toString().replace(":", "-");

        return temp + "_" + dateStr;
    }

}`

Base test class:
`public class BaseTestClass {
    public static ExtentReportUtil logger;
    public TestUtils testUtilObj = null;
    private static AndroidDriver driver;
    private AppiumDriverLocalService service;
    private AppiumServiceBuilder server;

    String platformRunAs;
    /*
     * myPlatform should be web or mobile
     *
     */
    final String myPlatform = "web";
    public static int portNumber;

    @Parameters({"platform","deviceName","port"})
    @BeforeClass
    public void setup(@Optional(myPlatform)String platform, @Optional("chrome_normal")String deviceName, int port) throws MalformedURLException
    {
        System.out.println(platform);
        if ("mobile".equals(platform))
        {
            System.out.println("Application is opening... ");
            DesiredCapabilities cap = new DesiredCapabilities();
            cap.setCapability("automationName","UiAutomator2");
            cap.setCapability("platformName", "Android");
            cap.setCapability("appPackage", "***************");
            cap.setCapability("appActivity", "*******************");
            cap.setCapability(MobileCapabilityType.NO_RESET,true);
            cap.setCapability("platformVersion","12");

            portNumber=port;
            if (deviceName.equals("Pixel_5_Bala"))
            {
                cap.setCapability("avd", deviceName);
            else if (deviceName.equals("Pixel_4_Bala"))
            {
                cap.setCapability("avd", deviceName);
            }

            ThreadLocalDriver.setTLDriver(new AndroidDriver(new URL("http://127.0.0.1:"+portNumber+"/wd/hub"), cap));
            driver = ThreadLocalDriver.getTLDriver();

            System.out.println("mobile will launch");
        }
        else
        {
            System.out.println("Incorrect Platform...");
        }

        ExtentReportUtil.createSparkReport();
        logger = new ExtentReportUtil();

    }

    @BeforeMethod
    public void invokeDriver(ITestContext context, ITestResult result, Method method){
        logger.createTest(method.getName());
        logger.createTestCategory(method.getDeclaringClass().getSimpleName());

        testUtilObj = new TestUtils(ThreadLocalDriver.getTLDriver(), logger);

    }
    @BeforeSuite
    public void setup(ITestContext ctx)
    {
        TestRunner runner = (TestRunner) ctx;
        runner.setOutputDirectory(System.getProperty("user.dir") + "/test-output");

    }

    @AfterSuite
    public void tearDown()
    {
        ExtentReportUtil.closeSparkReport();
        driver.quit();

    }

    public WebDriver getDriver()
    {
        return this.driver;
    }

    @Parameters({"runAs"})
    @AfterTest
    public void stopAppiumServer(@Optional(myPlatform) String platform)
    {
        if(platform.equalsIgnoreCase("mobile"))
        {
            service.stop();
            System.out.println("Appium Server is Stopped...");
        }
    }
}`
` <suite name="Test Suite" thread-count="2" parallel="tests" verbose="2">

    <test name="Pixel_5_Bala">
        <parameter name="runAs" value="grid"/>
        <parameter name="platform" value="mobile"/>
        <parameter name="deviceName" value="Pixel_5_Bala"/>
        <parameter name="port" value="4740"/>
        <classes>

            <class name="tests.LauncherTests"/>
        </classes>
    </test>

    <test name="Pixel_4_Bala">
        <parameter name="runAs" value="grid"/>
        <parameter name="platform" value="mobile"/>
        <parameter name="deviceName" value="Pixel_4_Bala"/>
        <parameter name="port" value="4760"/>
        <classes>
            <class name="tests.Signin"/>
        </classes>
    </test>
</suite>`

**Comment by Balakumar-14** [comment]
@jlipps @KazuCocoa Could you help me with this?

**Reply by jlipps** [reply]
I usually recommend running one appium server per device that way appium logs for parallel sessions are not interleaved

**Reply by Balakumar-14** [reply]
Thank you @jlipps. Here, I am trying to use grid and hub servers for both emulators. So, is it is possible to work parallel execution with grid and hub servers?
